### DISCUSSION TITLE
GUI styles

### DISCUSSION THREAD

**Discussion by kybos** [muse-sequencer/muse#1173] [Team Posts]
I'm trying to fix some issues with our GUI styles (like SDI, MDI, Borland), but it seems quite like never-ending story. Too many problems/bugs, and too many possible combinations of parameters to cope with.

I myself normally use the MDI/Cakewalk style, but in a very restricted manner, otherwise it invariably leads to a crash (e.g. using the tiling options from the Windows menu).

To make it short, I propose:

1. Remove all the individual options from the Settings->GUI styles tab and keep only the three predefined styles for user selection (SDI, MDI, Borland). This would at least give some predictability and defined states and avoid all the possible crazy combinations of parameters the user can do now.

2. (optional) Remove the GUI styles tab from Settings (there are too many tabs anyway already) and put the three fixed GUI styles as a dropdown somewhere else (maybe into the Styles tab of the Appearance dialog).

What do you think?

**Comment by terminator356** [comment]
I have no problem with that.
I suppose if it is possible to even keep the whole concept of styles is a challenge,
 at least we can honour @Windfisch 's vision by simplifying it.
I'd say do what you gotta do.
Maybe the crashes you are seeing can be fixed. I never saw any with these styles,
 with tiling and so on.

**Comment by terminator356** [comment]
Wow, I just noticed MDI support seems weak on Mint.
MDI windows have no minimize or maximize buttons.
When they are maximized, there are no buttons in the parent window to manipulate them,
 like there are in KDE. Only options were to use our tile/cascade commands.

**Comment by terminator356** [comment]
This is strange, in a simple 'stock' example MDI ui app they do have the buttons !
If I recall there are a couple of 'modes' the MDI windows can have, maybe that's why?

And a wee bug? I noticed that double-right-clicking on the title bar of a MusE MDI window
 makes it completely disappear, no way to retrieve it, Arranger not even listed in Window menu,
 tile/cascade greyed out. Weird.

**Comment by kybos** [comment]
Sorry, the missing buttons and menu in arranger are my change. I noticed too late that it’s more complicated than I thought. I will take it back until the full solution is available, as above.

**Comment by kybos** [comment]
I reverted all my related changes so you can check now.
I just got another crash, after minimizing a pianoroll and then tiling (in Cakewalk mode).

IMO the whole concept is very problematic, a weird mix of MDI and SDI, I never saw anything like that in another program (the GIMP one-window mode is presumably not a classic MDI, rather a docking concept, which is the preferred layout with most newer software nowadays). 
But I guess we have to keep the three GUI styles, as we would probably never agree which is the "right one" to keep ;-).

**Comment by terminator356** [comment]
Good day. Notice that Qt Designer has two modes, MDI and docking mode.
I asked this question once: I wonder what our 'Borland/Mac' mode would like like on an _actual_ Mac !
I must ask my employer if I can borrow one. I know Robert has worked on it before, we have a branch.

Also, for the first time in my life (don't laugh) I have both a donated iPhone - and a donated Android tablet !
Guess which one I'm playing with. Yeah, I am very curious if... it's... possible on the tablet.

**Comment by terminator356** [comment]
I should point out that a few people have asked for 'docked' windows, or 'docking'.
I am a bit confused. What Qt Designer calls 'Docked' appears to be MDI,
 while the unwieldy 'Multiple Top-Level Windows' is like the Borland/Mac style.
And meanwhile the API docs refer to dock areas - only on the top, left, bottom, and right,
 but I guess those are specifically _toolbar_ dock areas.
So I had assumed that what these users meant was some kind of windows that 'stick together' (?).
Because we already have MDI and MTLW, so what could they possibly mean by docked?
[ Edit ] Sorry, I see. Dockable is _like_ MDI but the windows can be 'detached' and 'reattached'.
Right, like Qt Designer's windows. Dockable. But similar to MDI.

BTW Each and every menu can be easily made 'tear-off'.
One of ours, I think in PianoRoll, is set like that for some reason.

**Comment by kybos** [comment]
Yes, it's confusing, the terminology is not clear (dockable windows or menus or whatever...). 

What I meant with the strange concept in MusE is not that it supports different GUI modes, but that it mixes them - you will most probably never find something like that in any other program.
Just take the default "SDI" style. It's a mix of
1. MDI with subwindow for the Arranger window.
2. SDI for all the other top windows (like MIDI editors).
3. Lot of special handling for the Arranger window.
The arranger IS special indeed and should be handled as such, but then in a complete and consistent way (that's what I had started to implement, based on a user ticket).

As for the state-of-the-art one-window-concept with docks: I don't think it's peculiarly suitable for MusE, and it would probably mean rewriting most of the GUI. Not realistic IMO. The GUI modes we have offer enough flexibility.

BTW: Qt's support for the classic MDI mode is quite bad. You can see it in our Cakewalk mode - the title bars of the subwindows are not styled properly (and cannot be), they are always blue, with blurred texts. There are many complaints from users about that, but Qt people do not seem to care.

> BTW Each and every menu can be easily made 'tear-off'.
> One of ours, I think in PianoRoll, is set like that for some reason.

Yes, the Functions menu. I noticed that of course, as I am very sensitive to this kind of UI inconsistencies ;-). Just one menu of this kind looks like a mistake or an experiment, which is not good.
The tearable menus are nothing new, but they are not widely used, as far as I can judge.

**Comment by terminator356** [comment]
I am grateful we have someone dedicated to this kind of thing,
 who is now intimately familiar with the code. ("Familiarity breeds contempt" ha ha).
Robert and I can do a lot of UI stuff but when it comes to mass migration
 towards some new style, I don't think we have the time, we just want to make it work.
People have stopped by to offer all kinds of suggestions or demands,
 but rarely understand the code and difficulties in implementing the suggestions.

If you have a vision of what it could be, I think I would trust it ;-)
What do you think would be the best way forward?
Or is it sort of "one step at a time, let's see what we have and how to improve it" kind of thing?

**Comment by Windfisch** [comment]
To be frank, even the MDI style was only a dirty kludge to reduce the waste of screen space similarly to what early cakewalk versions did.
In the long run it might be more profitable to completely throw away all three GUI styles and to convert MusE into a dockable-panels-like style. What's most important about this is:
MusE needs unified menus: Every window or pane should have exactly the same toolbar and menu entries. Some might be greyed out in certain situations. Functionality that is not song-specific but really editor-specific might be placed in a (small!) popuo menu button in the corresponding pane.
Just my 2c

**Comment by kybos** [comment]
[Please consider that I didn't see Florian's text above when writing this...]

You would really go for it? ;-)

In that case: Yes, the docking concept is the state-of-the-art GUI mode, it's kind of synthesis of SDI and MDI with even more features. I even think it's something Florian was trying to achieve with his mode mix, but the proper technology was not there yet at that time. 

My vision would be (in abstract terms): Simplify where possible, keep the UI consistent and predictable, make the UI as customizable as possible.

In concrete steps:
1. Replace the current GUI styles with a clean docking concept.
2. Qt styles: Stick to Fusion as basis for our own themes. Most non-Qt style C++ plugins (Breeze/Oxygen etc.) don't support external css styling well, so the behaviour is unpredictable and often not as intended by the theme author.
3. Non-standard buttons in the mixer strips (and the transport), like Mute, Record, Input etc. Technically these are not (Qt) buttons at all, they don't look like buttons and are not customizable/themable in any way. Also the variable placement of the buttons at the bottom of the mixer strip for different track types is an issue to be solved IMO.
